app/Modules/Auth/AuthController.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Modules\Auth;
+
+use App\Modules\Auth\Models\User;
+use Tempest\Auth\Authenticator;
+use Tempest\Http\Get;
+use Tempest\Http\Post;
+use Tempest\Http\Request;
+use Tempest\Http\Response;
+use function Tempest\response;
+
+final readonly class AuthController
+{
+    public function __construct(
+        private Authenticator $authenticator,
+    ) {
+    }
+
+    #[Post('/login')]
+    public function login(Request $request): Response
+    {
+        $user = (new User(name: 'John Doe'))->setCredentials(
```

Comment:
I think it's weird that we create a new User object, why don't we retrieve it from the database?

Reply:
I understand that we're doing the retrieving somewhere else, but imo this is a code smell. Either we should do the retrieval in the controller, and `Authenticator::login` only sets up the session; or we should do it in `DatabaseAuthenticator`, which means the `login` signature should change to `Authenticator::login($name, $password)`

IMO, the first approach is better, because there will be ways we can log in identifiables without username/password (think oauth, login links, …)

So, I would say the `Authenticator`'s concern is only about managing the login session for a user. It'll be interesting to brainstorm ideas about being able to log in outside a web context (CLI), and to log in as an admin user acting as another user. Those are concerns that the authenticator should be able to handle, but it shouldn't care about where an identifiable came from.

I think that leads us to conclude that we need something else that can resolve an identifiable from some kind of user input (username/password, login link, oauth token, …). I think there's a gap to be filled there as well, but maybe it's better to first have this merged, and then work on the resolving part separately.

WDYT?

Reply:
That's true!
I think these should be two separate things.
Here's an idea:

1 - We update the `Identifiable` to have just one method: `identifier` that's going to be the unique identifier for it

2 - We create another interface called `CredentialsResolver` or something like this. This is going to be an empty interface and we would use it only to mark a class as the one used to resolve some credentials. Why should it be empty? Because getting credentials for web, CLI, OAuth, etc, are all different. Then we could use this to have implementations for different types of credentials, for example, the Database one would be something like this.

```php
final readonly class DatabaseCredentials implements CredentialsResolver
{
    public function __construct(
        private string $source,
        private string $identifierField,
        private string $secretField,
    ) {
    }

    public function resolve(string $identifier, string $password): Identifiable
    {
        // Get user from DB here
    }
}
```

For a login it would be something like this:

```php
final readonly class AuthController
{
    public function __construct(
        private CredentialsResolver $credentialsResolver,
        private Authenticator $authenticator,
    ) {
    }

    #[Post('/login')]
    public function login(Request $request): Response
    {
        $user = $this->credentialsResolver->resolve(
            identifier: $request->get('email'),
            password: $request->get('password')
        );

        $this->authenticator->login($user);

        return response()->ok();
    }
}
```

The only thing I'm not sure for this approach is the best way to configure the three parameters of the `DatabaseCredentials`

Reply:
> Why should it be empty? Because getting credentials for web, CLI, OAuth, etc, are all different

I was thinking about this as well. Another approach is to pass the variable data to the constructor of each resolver, and have the method be something like this: `resolve(): Identifiable`.

If the constructor is already saturated with input and we don't want to add any more params (also because we might want to register it in the container etc), we could have some kind of object and do manual instanceof checks

```php
final readonly class DatabaseCredentials implements CredentialsResolver
{
    public function __construct(
        private string $source,
        private string $identifierField,
        private string $secretField,
    ) {}

    public function resolve(AuthenticationRequest $request): Identifiable
    {
        if (! $request instanceof DatabaseAuthenticationRequest) {
            throw new Exception;
        }
        
        // …
    }
}
```

Just FYI, these are the kinds of problems for which I wish we had generics 😁

Reply:
Also: `AuthenticationRequest` is a sucky name that'll need to be changed

Reply:
I like the idea, I'm going to make some changes in the PR to follow this approach!
I think that generics would be a HUGE improvement to the language.